File: 올마나마/scripts/coupang_api.py
# -*- coding: utf-8 -*-
"""쿠팡 파트너스 API — **부작용 없는 공용 핵심**. 2026-07-31 신설

    from coupang_api import load_env, search, keys
    ak, sk = keys()
    body = search("동치미 냉면육수", ak, sk)

🔴 왜 갈라냈나 (2026-07-31 실제 사고)
   `coupang_search.py` 는 파일 맨 끝에 **`sys.exit(main())`** 이 있다. 그래서 다른 데서
   `import coupang_search` 만 해도 그 자리에서 프로세스가 죽는다.
   관리자 서버(admin_rc.py)에서 「쿠팡 찾기」를 눌렀더니 **서버가 통째로 죽었다.**
   → 부르는 쪽이 늘어날수록 위험하니 «불러 쓰는 부분»만 여기로 옮겼다.
   `coupang_search.py` 는 CLI 로 그대로 둔다(이미 잘 돌고 있다).

⚠️ **분당 50회.** 여러 번 부를 때는 부르는 쪽이 `time.sleep(1.3)` 을 넣는다(분당 46회).
⚠️ 이 API 는 실패해도 HTTP 200 을 준다 — `rCode` 를 반드시 본다. `ok()` 가 그걸 한다.
"""
import hashlib
import hmac
import io
import json
import os
import re
import ssl
import threading
import time
import urllib.parse
import urllib.request
from datetime import datetime, timezone

BASE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DOMAIN = "https://api-gateway.coupang.com"
PATH = "/v2/providers/affiliate_open_api/apis/openapi/v1/products/search"
# 🔴 최대 10. 20 을 주면 rCode=400 «limit is out of range» 다(2026-07-31 확인).
LIMIT = 10


def load_env():
    env = {}
    p = os.path.join(BASE, ".env")
    if os.path.exists(p):
        for line in io.open(p, encoding="utf-8-sig"):
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                env[k.strip()] = v.strip()
    for k in ("COUPANG_ACCESS_KEY", "COUPANG_SECRET_KEY"):
        if os.environ.get(k):
            env[k] = os.environ[k]
    return env


def keys():
    e = load_env()
    return e.get("COUPANG_ACCESS_KEY", ""), e.get("COUPANG_SECRET_KEY", "")


def sign(method, url_path, secret, access):
    """CEA HMAC-SHA256. signed-date 는 **GMT** 여야 한다 — 로컬시간 쓰면 401."""
    parts = url_path.split("?")
    path, query = parts[0], (parts[1] if len(parts) > 1 else "")
    dt = datetime.now(timezone.utc).strftime("%y%m%dT%H%M%SZ")
    msg = dt + method + path + query
    sig = hmac.new(secret.encode(), msg.encode(), hashlib.sha256).hexdigest()
    return ("CEA algorithm=HmacSHA256, access-key=%s, signed-date=%s, signature=%s"
            % (access, dt, sig))


# 🔴 **분당 50회를 코드로 막는다** (2026-07-31 신설).
#    2026-07-27 에 3초 간격으로 몰아 읽다 11건에서 차단당했다. 부르는 쪽이 sleep 을
#    넣어주길 기대하지 않는다 — 여기서 지킨다. 분당 46회(1.3초)로 여유를 둔다.
RATE_GAP = 1.3
_rate_lock = threading.Lock()
_last_call = [0.0]


def _throttle():
    with _rate_lock:
        wait = RATE_GAP - (time.monotonic() - _last_call[0])
        if wait > 0:
            time.sleep(wait)
        _last_call[0] = time.monotonic()


def search(keyword, ak, sk, limit=LIMIT):
    _throttle()
    url_path = "%s?keyword=%s&limit=%d" % (PATH, urllib.parse.quote(keyword), limit)
    req = urllib.request.Request(DOMAIN + url_path, method="GET")
    req.add_header("Authorization", sign("GET", url_path, sk, ak))
    req.add_header("Content-Type", "application/json;charset=UTF-8")
    with urllib.request.urlopen(req, timeout=25,
                                context=ssl.create_default_context()) as r:
        return json.loads(r.read().decode("utf-8"))


# 🔴 쿠팡 파트너스 **필수 고지** — 링크를 싣는 페이지에 이 문구가 있어야 한다.
#    API 응답 rMessage 로도 계속 내려온다. 계산기 페이지 하단에 반드시 노출할 것.
DISCLOSURE = "이 포스팅은 쿠팡 파트너스 활동의 일환으로, 이에 따른 일정액의 수수료를 제공받습니다."


RX_PACK = re.compile(r"(\d+(?:\.\d+)?)\s*(kg|g|ml|l|리터)(?![a-z가-힣])", re.I)
# 🔴 「개입」·「개들이」를 먼저 놓는다 (2026-07-31 공통모듈로 옮기며 발견한 버그).
#    전에는 `개|입|…` 뿐이라 **「500g 2개입」의 배수를 통째로 놓쳤다**(500g 로 읽혔다).
#    「2개」 까지 맞춘 뒤 뒤따르는 「입」이 `(?![가-힣])` 에 걸려 매칭이 통째로 취소된 것이다.
RX_CNT = re.compile(
    r"(?<![\d.])(\d{1,3})\s*(?:개입|개들이|개|입|팩|봉|포|구|알|장|병|캔)(?![가-힣])")


def parse_pack(title):
    """상품명에서 **총 용량**을 읽는다 → (양, 단위). 못 읽으면 (None, None).

    🔴 **배수를 반드시 곱한다.** 식봄에서 「5kg 2팩」 의 배수를 놓쳐 562건이 틀렸다(2026-07-29).
       쿠팡은 `냉면육수, 20개, 310ml` 처럼 **개수가 용량 앞에 오기도** 한다.

    ⛔ 개수가 여러 개 나오면(예: `(4인분)x3개, 1.816kg, 3개`) **추측하지 않고 None** 을 준다.
       틀린 용량은 「몇 인분 커버」 를 통째로 망가뜨린다 — 빈칸이 가짜보다 낫다.
    """
    tt = title or ""
    # 🔴 **값이 아니라 «몇 번 나왔나»로 센다** (2026-07-31 고침).
    #    전에는 집합(set)으로 모아 값이 같으면 하나로 합쳐졌다. 그래서 아래 주석이
    #    «None 이어야 한다» 고 적어둔 `(4인분)x3개, 1.816kg, 3개` 가 5,448g 으로 나왔다.
    hits = [int(x) for x in RX_CNT.findall(tt) if 1 < int(x) <= 200]
    m = RX_PACK.search(tt)
    if not m:
        # 무게·부피가 없고 **개수만** 있는 것 — 계란 10구, 오이 3입.
        # 이런 재료는 필요량도 「개」 로 적으므로 그대로 쓰면 된다.
        if len(hits) == 1:
            return float(hits[0]), "개"
        return None, None
    v, u = float(m.group(1)), m.group(2).lower()
    if u == "kg":
        v, u = v * 1000, "g"
    elif u in ("l", "리터"):
        v, u = v * 1000, "ml"
    if len(hits) > 1:
        return None, None            # 개수 표기가 여럿이다 — 사람이 봐야 한다
    if hits:
        v *= hits[0]
    return round(v, 1), u


DEEPLINK_PATH = "/v2/providers/affiliate_open_api/apis/openapi/v1/deeplink"


def deeplink(urls, ak, sk):
    """쿠팡 상품 주소 → **제휴가 붙은 주소**. 2026-07-31 확인.

    🔴 왜 필요한가 — 검색 API 는 **상위 10개**만 주고 호출마다 결과가 바뀐다.
       웹에서 눈으로 찾은 더 싼 상품을 쓰려면 그 주소를 제휴 링크로 바꿔야 한다.
       그냥 상품 주소를 걸면 **수수료가 안 붙는다.**

    응답 `data:[{originalUrl, shortenUrl, landingUrl}]` — 짧은 건 `shortenUrl`.
    """
    _throttle()
    if isinstance(urls, str):
        urls = [urls]
    body = json.dumps({"coupangUrls": urls}).encode()
    req = urllib.request.Request(DOMAIN + DEEPLINK_PATH, data=body, method="POST")
    req.add_header("Authorization", sign("POST", DEEPLINK_PATH, sk, ak))
    req.add_header("Content-Type", "application/json;charset=UTF-8")
    with urllib.request.urlopen(req, timeout=25,
                                context=ssl.create_default_context()) as r:
        return json.loads(r.read().decode("utf-8"))


def product_id_of(url):
    """주소에서 상품번호를 뽑는다 — `/vp/products/2233944921?...` → 2233944921"""
    m = re.search(r"/products/(\d+)", url or "")
    return m.group(1) if m else None


def item_id_of(url):
    """옵션 식별자. 같은 상품(productId)이라도 **옵션마다 itemId 가 다르다.**"""
    m = re.search(r"itemId=(\d+)", url or "")
    return m.group(1) if m else None


def group_options(items):
    """같은 productId 를 **옵션 묶음**으로 표시한다. 2026-07-31 신설

    🔴 왜 — 「면사랑 함흥면사리(2kg)」 가 1,250원·5,200원 **두 줄**로 왔다. productId 가 같다.
       상품명은 옵션 하나의 이름일 뿐이라 **1,250원이 2kg 값이라는 근거가 없다.**
       (같은 목록의 「10인분 9,900원」 = 1인분 990원과 견주면 1인분 104원은 말이 안 된다)

    ⛔ 그렇다고 버리지 않는다 — 옵션이 곧 «작은 용량» 인 경우가 많다.
       옵션을 전부 살려두고 **몇 개짜리인지 사람이 보고 고르게** 한다.
    """
    seen = {}
    for x in items:
        seen.setdefault(str(x.get("product_id") or x.get("productId") or ""), []).append(x)
    for x in items:
        pid = str(x.get("product_id") or x.get("productId") or "")
        x["opt_n"] = len(seen.get(pid, []))
    return items


# 로켓프레시는 별도 칸이 없다 — **상품명에 붙어서** 온다(2026-07-31 확인).
#    예: `[로켓프레시] 면사랑 함흥면사리 10인분 (냉동)`
RX_FRESH = re.compile(r"로켓프레시|로켓\s*프레시", re.I)


def is_fresh(title):
    return bool(RX_FRESH.search(title or ""))


def ok(body):
    """rCode 0 이 아니면 실패다 — 200 이라고 믿으면 «0건 정상» 으로 착각한다."""
    return str((body or {}).get("rCode")) == "0"


def items(body):
    return ((body or {}).get("data") or {}).get("productData") or []

File: sajangpick/scripts/coupang-api.mjs
// 쿠팡 파트너스 Open API 연동 모듈 — 골격만 만들어 둔 상태.
//
// 왜 지금 만들어 두나: 파트너스 최종승인(누적 판매 15만원)이 나면 키가 생기는데,
// 그때 키만 넣으면 바로 돌아가게 하려는 것. 지금은 네트워크 호출을 하지 않는다.
//
// 키 넣는 법: 프로젝트 루트에 .env 파일을 만들고 아래 두 줄을 적는다.
//   COUPANG_ACCESS_KEY=...
//   COUPANG_SECRET_KEY=...
// .env는 git에 올라가지 않는다(.gitignore에 등록됨).
//
// ⚠ 엔드포인트 경로는 공식 문서에서 확인한 값으로 채워야 한다.
//    추측해서 호출하면 안 된다(경로를 무작위로 찔러보면 차단 대상이 된다).
//    키를 받는 시점에 파트너스 개발자 문서를 열어 PATHS를 확정할 것.

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

const HOST = 'https://api-gateway.coupang.com';

// 문서 확인 후 확정할 것. 지금 값은 자리표시자다.
const PATHS = {
  search: '/v2/providers/affiliate_open_api/apis/openapi/products/search',
  deeplink: '/v2/providers/affiliate_open_api/apis/openapi/v1/deeplink',
};

// .env를 아주 단순하게 읽는다(라이브러리 추가 없이).
function loadEnv() {
  const file = path.join(ROOT, '.env');
  if (!fs.existsSync(file)) return {};
  const out = {};
  for (const line of fs.readFileSync(file, 'utf8').split(/\r?\n/)) {
    const m = line.match(/^\s*([A-Z0-9_]+)\s*=\s*(.*)\s*$/);
    if (m) out[m[1]] = m[2];
  }
  return out;
}

export function getKeys() {
  const env = { ...loadEnv(), ...process.env };
  const accessKey = env.COUPANG_ACCESS_KEY;
  const secretKey = env.COUPANG_SECRET_KEY;
  return { accessKey, secretKey, ready: Boolean(accessKey && secretKey) };
}

// 쿠팡은 요청마다 HMAC 서명을 요구한다.
// 서명 대상 문자열 = 시각 + HTTP메서드 + 경로 + 쿼리스트링
function buildAuthHeader({ method, urlPath, query = '', accessKey, secretKey }) {
  // yyMMddTHHmmssZ (GMT)
  const signedDate = new Date().toISOString().replace(/[-:]/g, '').replace(/\..+/, 'Z').slice(2);
  const message = signedDate + method + urlPath + query;
  const signature = crypto.createHmac('sha256', secretKey).update(message).digest('hex');
  return `CEA algorithm=HmacSHA256, access-key=${accessKey}, signed-date=${signedDate}, signature=${signature}`;
}

async function request({ method, urlPath, query = '', body = null }) {
  const { accessKey, secretKey, ready } = getKeys();
  if (!ready) {
    throw new Error('쿠팡 파트너스 API 키가 없습니다 (.env의 COUPANG_ACCESS_KEY / COUPANG_SECRET_KEY)');
  }

  const auth = buildAuthHeader({ method, urlPath, query, accessKey, secretKey });
  const res = await fetch(`${HOST}${urlPath}${query ? `?${query}` : ''}`, {
    method,
    headers: { Authorization: auth, 'Content-Type': 'application/json' },
    body: body ? JSON.stringify(body) : undefined,
  });

  if (!res.ok) throw new Error(`쿠팡 API ${res.status}: ${(await res.text()).slice(0, 200)}`);
  return res.json();
}

// 상품 검색 → 이미지 URL 확보용.
// 반환 형태는 실제 응답을 보고 맞춰야 하므로 지금은 원본을 그대로 넘긴다.
export async function searchProduct(keyword, limit = 5) {
  const query = `keyword=${encodeURIComponent(keyword)}&limit=${limit}`;
  return request({ method: 'GET', urlPath: PATHS.search, query });
}

// 상품 URL을 파트너스 추적 링크로 바꾼다.
export async function createDeeplink(urls) {
  return request({ method: 'POST', urlPath: PATHS.deeplink, body: { coupangUrls: urls } });
}

// 렌더 파이프라인이 부르는 자리.
// 키가 없으면 이미지 없이 진행하라고 알려주고 넘어간다(렌더를 막지 않는다).
export async function fetchProductImages(plan) {
  const { ready } = getKeys();
  if (!ready) {
    return { ok: false, reason: 'API 키 미발급 — 이미지 없이 렌더', images: [] };
  }
  const result = await searchProduct(plan.product_name);
  return { ok: true, raw: result, images: [] }; // 응답 구조 확인 후 images 매핑을 채울 것
}
